blockstore/net.go

```diff
@@ -0,0 +1,462 @@
+package blockstore
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/libp2p/go-msgio"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+)
+
+type NetRPCReqType byte
+
+const (
+	NRpcHas     NetRPCReqType = iota
+	NRpcGet     NetRPCReqType = iota
+	NRpcGetSize NetRPCReqType = iota
+	NRpcPut     NetRPCReqType = iota
+	NRpcDelete  NetRPCReqType = iota
+	NRpcList    NetRPCReqType = iota
+
+	// todo cancel req
+)
+
+type NetRPCRespType byte
+
+const (
+	NRpcOK   NetRPCRespType = iota
+	NRpcErr  NetRPCRespType = iota
+	NRpcMore NetRPCRespType = iota
+)
+
+type NetRPCErrType byte
+
+const (
+	NRpcErrGeneric  NetRPCErrType = iota
+	NRpcErrNotFound NetRPCErrType = iota
+)
+
+type NetRpcReq struct {
+	Type NetRPCReqType
+	ID   uint64
+
+	Cid  []cid.Cid // todo maxsize?
```

Comment:
Yeah, generally this side is just used for serving requests made by lotus, which should hopefully behave nicely.

I'll add a note on security in case anyone tries to re-use this code outside of lotus.